stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+                          Value inputVal, double inputScale, int64_t inputZp) {
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
+  if (!inputType) {
+    return nullptr;
+  }
+  auto outputType = inputType.clone(rewriter.getI32Type());
+
+  const int32_t scaleWidth = 32;
+  int32_t multiplier, shift;
+  computeMultiplierAndShift(inputScale, multiplier, shift, scaleWidth);
+
+  return buildRescale(rewriter, loc, outputType, inputVal, multiplier, shift,
+                      inputZp,
+                      /* output_zp = */ 0,
+                      /* double_round = */ false,
+                      /* scale32 = */ true,
+                      /* per_channel = */ false);
+}
+
+// Creates TOSA rescale op with int32 input
+Value buildRescaleFromInt32(PatternRewriter& rewriter, Location loc,
+                            mlir::ShapedType outputType, Value inputVal,
+                            double outputScale, int64_t outputZp) {
+  // Input should be int32 type
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
+  if (!inputType || !inputType.getElementType().isInteger(32)) {
+    // expected rescale input element type to be i32
+    return nullptr;
+  }
+
+  const int32_t scaleWidth = 32;
+  int32_t multiplier, shift;
+  computeMultiplierAndShift(outputScale, multiplier, shift, scaleWidth);
+
+  return buildRescale(rewriter, loc, outputType, inputVal, multiplier, shift,
+                      /* input_zp = */ 0, outputZp,
+                      /* double_round = */ false,
+                      /* scale32 = */ true,
+                      /* per_channel = */ false);
+}
+
+template <typename StablehloOp>
+LogicalResult matchAndRewriteAddSub(StablehloOp op, PatternRewriter& rewriter) {
+  Value lhs = op.getLhs();
+  Value rhs = op.getRhs();
+  Value result = op.getResult();
+
+  auto lhsType = dyn_cast<mlir::ShapedType>(lhs.getType());
+  auto rhsType = dyn_cast<mlir::ShapedType>(rhs.getType());
+  auto resultType = dyn_cast<mlir::ShapedType>(result.getType());
+
+  if (!lhsType || !rhsType || !resultType) {
+    return rewriter.notifyMatchFailure(
+        op, "inputs and result tensors should all have shaped tensor types");
+  }
+
+  auto lhsQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(lhsType.getElementType());
+  auto rhsQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(rhsType.getElementType());
+  auto resultQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(resultType.getElementType());
+
+  if (!lhsQType || !rhsQType || !resultQType) {
+    if (lhsQType || rhsQType || resultQType) {
+      return rewriter.notifyMatchFailure(op,
+                                         "inputs and result tensors should be "
+                                         "all quantized or all un-quantized");
+    }
+    // not quantized - skip
+    return failure();
+  }
+
+  // lhs/rhs/result are all quantized
+
+  // Following quantization described in tensorflow/lite/kernels/add.cc
```

Comment:
Lets avoid the source here just for the sake of maintaining the relevance of comment in the event of unrelated code refactoring.

Reply:
done